cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh

```diff
@@ -0,0 +1,488 @@
+                             const int k,
+                             const bool greater,
+                             const int pass)
+{
+  __shared__ bool isLastBlockDone;
+
+  constexpr int num_buckets = calc_num_buckets<BITS_PER_PASS>();
+  constexpr int num_passes  = calc_num_passes<T, BITS_PER_PASS>();
+  const int batch_id        = blockIdx.y;
+  in_buf += batch_id * len;
+  out_buf += batch_id * len;
+  if (in_idx_buf) { in_idx_buf += batch_id * len; }
+  if (out_idx_buf) { out_idx_buf += batch_id * len; }
+  if (out) {
+    out += batch_id * k;
+    out_idx += batch_id * k;
+  }
+  auto counter   = counters + batch_id;
+  auto histogram = histograms + batch_id * num_buckets;
+
+  filter_and_histogram<T, IdxT, BITS_PER_PASS>(in_buf,
+                                               in_idx_buf,
+                                               out_buf,
+                                               out_idx_buf,
+                                               out,
+                                               out_idx,
+                                               len,
+                                               counter,
+                                               histogram,
+                                               greater,
+                                               pass,
+                                               k);
+  __threadfence();
+
+  if (threadIdx.x == 0) {
+    unsigned int finished = atomicInc(&counter->finished_block_cnt, gridDim.x - 1);
+    isLastBlockDone       = (finished == (gridDim.x - 1));
+  }
+
+  // Synchronize to make sure that each thread reads the correct value of
+  // isLastBlockDone.
+  __syncthreads();
+  if (isLastBlockDone) {
+    if (counter->len == 1 && threadIdx.x == 0) {
+      counter->previous_len = 0;
+      counter->len          = 0;
+    }
+    // init counter, other members of counter is initialized with 0 by
+    // cudaMemset()
+    if (pass == 0 && threadIdx.x == 0) {
+      counter->k   = k;
+      counter->len = len;
+      if (out) { counter->out_back_cnt = 0; }
+    }
+    __syncthreads();
+
+    IdxT ori_k = counter->k;
+
+    if (counter->len > 0) {
+      choose_bucket<T, IdxT, BITS_PER_PASS, NUM_THREAD>(counter, histogram, ori_k);
+    }
+
+    __syncthreads();
+    if (pass == num_passes - 1) {
+      const IdxT previous_len = counter->previous_len;
+      const int want_bucket   = counter->bucket;
+      int start_bit           = calc_start_bit<T, BITS_PER_PASS>(pass);
+      unsigned mask           = calc_mask<T, BITS_PER_PASS>(pass);
+
+      if (!out) {  // radix select
+        for (IdxT i = threadIdx.x; i < previous_len; i += blockDim.x) {
+          const T value = out_buf[i];
+          int bucket    = calc_bucket<T, BITS_PER_PASS>(value, start_bit, mask, greater);
+          if (bucket == want_bucket) {
+            // TODO: UB
+            // could use atomicExch, but it's not defined for T=half
+            counter->kth_value = value;
```

Comment:
This `if` branch is for another top-k implementation called "radix-select", which only sets counter->kth_value during the passes without writing top-k values to the result array `out`.
After all passes are done, the counter->kth_value is used as a threshold to filter out the real top-k values from original inputs. So counter->kth_value is just an intermediate value and is never passed back to the caller.

Since this implementation is removed, I think this `if` branch is not used anymore and should be deleted.